// scirpus/lib/assert.js
"use strict";

var type = require("type-of");

var fmt = function fmt(x) {
  return JSON.stringify(x);
};

exports.string = function (x, comment) {
  if (type(x) !== "string") {
    throw new Error("expects string but got " + fmt(x) + " at " + comment);
    return;
  }

  return undefined;
};

exports.array = function (x, comment) {
  if (type(x) !== "array") {
    throw new Error("expects array but got " + fmt(x) + " at " + comment);
    return;
  }

  return undefined;
};

exports.func = function (x, comment) {
  if (type(x) !== "function") {
    throw new Error("expects function but got " + fmt(x) + " at " + comment);
    return;
  }

  return undefined;
};

exports.oneOf = function (x, xs, comment) {
  if (!(xs["indexOf"](x) >= 0)) {
    throw new Error(fmt(x) + " is not oneOf " + fmt(xs) + " at " + comment);
    return;
  }

  return undefined;
};

exports.result = function (x, comment) {
  if (!x) {
    throw new Error("expects true but got " + fmt(x) + " at " + comment);
    return;
  }

  return undefined;
};

exports.defined = function (x, comment) {
  if (!(x != null)) {
    throw new Error("value not defined at " + comment);
    return;
  }

  return undefined;
};
